web-common/src/features/entity-management/file-artifacts.ts

```diff
@@ -337,6 +351,35 @@ export class FileArtifacts {
       },
     );
   }
+
+  public getNamesForKind(kind: ResourceKind): string[] {
+    return (
+      this.artifactsByKind[kind]?.map((a) => get(a.name)?.name ?? "") ?? []
+    );
+  }
+
+  private newFileArtifact(filePath: string) {
+    const artifact = new FileArtifact(filePath, (prevName, name) =>
+      this.handleArtifactNameChange(artifact, prevName, name),
+    );
+    return artifact;
+  }
+
+  private handleArtifactNameChange(
+    fileArtifact: FileArtifact,
+    prevName: V1ResourceName | undefined,
+    name: V1ResourceName | undefined,
+  ) {
+    if (prevName?.kind && this.artifactsByKind[prevName.kind]) {
+      const index = this.artifactsByKind[prevName.kind].find(
+        (a: FileArtifact) => a.path === fileArtifact.path,
+      );
+      if (index >= 0) this.artifactsByKind[prevName.kind].splice(index, 1);
+    }
+    if (!name?.kind) return;
+    this.artifactsByKind[name.kind] ??= [];
+    this.artifactsByKind[name.kind].push(fileArtifact);
+  }
 }
```

Comment:
This `handleArtifactNameChange` callback is quite confusing – see comment and suggestion above.